server/events.py

```diff
@@ -0,0 +1,98 @@
+from __future__ import annotations
+
+from enum import Enum
+from typing import Literal, TypedDict
+
+from pydantic import BaseModel
+
+Position = tuple[int, int]
+Replacement = TypedDict("Replacement", {"from": int, "to": int, "value": str})
+
+
+class EventType(Enum):
+    """The type of a WebSocket event."""
+
+    CONNECT = "connect"
+    DISCONNECT = "disconnect"
+    SYNC = "sync"
+    MOVE = "move"
+    REPLACE = "replace"
+
+
+class EventData(BaseModel):
+    """The data of a WebSocket event.
+
+    This is just a base class that other classes will inherit from.
+    """
+
+
+class ConnectionData(EventData):
+    """The data of a connection event.
+
+    Fields:
+        connection_type: "create" if the user wants to create the room, "join"
+            if the user wants to join the room.
+        difficulty (optional): The difficulty of the room, only needed if the
+            "connection_type" is "create".
+        room_code: The unique four-letters code that will represent the room.
+        username: The username of the user creating or joining the room.
+    """
+
+    connection_type: Literal["create", "join"]
+    difficulty: int | None
+    room_code: str
+    username: str
+
+
+class DisconnectData(EventData):
+    """The data of a disconnection event.
+
+    Fields:
+        room_code: The code of the room from which the user wants to disconnect.
+    """
+
+    room_code: str
+
+
+class SyncData(EventData):
+    """The data of a sync event.
+
+    Fields:
+        code: The code that already exists in the room.
+        collaborators: The list of users that already collaborate in the room.
+    """
+
+    code: str
+    collaborators: list[str]
```

Comment:
Does this not need the `room_code` as well?

Reply:
It shouldn't be needed. The `room_code` is obtained with the initial `connection` event, otherwise, every event would need the room code.

But now that you make me think about that, it probably isn't even needed in the `DisconnectData`.

Reply:
Yeah, I agree. Don't think it needs to be in `DisconnectData`.

Reply:
But it probably needs `username`, so that the client knows which collaborator to remove from the list.

Reply:
Agreed.